File: aaaaaa/emelixir/lib/buffer.ex
defmodule Buffer do
  def init do
    IO.puts("Buffer created.")
    max_size = 10
    buffer_array = []
    item_id_prod = 1
    listen(max_size, buffer_array, item_id_prod)
  end

  defp listen(max_size, buffer_array, item_id_prod) do
    receive do
      {:available, from, item, prod_id} when length(buffer_array) < max_size ->
        new_buffer_array = buffer_array ++ [{item_id_prod, item}]
        send(from, :put_item)
        IO.puts("Producer #{prod_id} inserted item #{item_id_prod}. There are #{length(new_buffer_array)} items in buffer.")
        output(new_buffer_array)
        listen(max_size, new_buffer_array, item_id_prod + 1)

      {:occupied, from, cons_id} when buffer_array != [] ->
        [{item_id, item} | new_buffer_array] = buffer_array
        send(from, :remove_item)
        IO.puts("Consumer #{cons_id} removed item #{item_id}. There are #{length(new_buffer_array)} items in buffer.")
        output(new_buffer_array)
        listen(max_size, new_buffer_array, item_id_prod)

      {:available, from, _item, _prod_id} ->
        send(from, :full)
        listen(max_size, buffer_array, item_id_prod)

      {:stop} ->
        IO.puts("Buffer stopping.")
    end
  end

  defp output(buffer) do
    item_ids = Enum.map(buffer, &elem(&1, 0))
    IO.puts("Buffer: [#{Enum.join(item_ids, ", ")}]")
  end

  defp fnl([h]) do
    IO.puts("#{get_item_id(h)}]")
  end

  defp fnl([h | t]) do
    IO.write("#{get_item_id(h)}, ")
    fnl(t)
  end

  defp fnl([]) do
    IO.puts("]")
  end

  defp get_item_id({item_id, _item}), do: item_id
end

File: aaaaaa/#Entrega/buffer.erl
-module(buffer).
-export([init/0, listen/4, output/1, getItemId/1]).

init() ->
  io:format("Buffer criado~n"),
  MaxSize = 40,
  BufferArray = [],
  ItemId_Prod = 1,
  listen(MaxSize, BufferArray, ItemId_Prod, 0).

listen(MaxSize, BufferArray, ItemId_Prod, Type) ->
  receive
    {available, From, Item, Prod_Id, ItemType} ->
      case length(BufferArray) < MaxSize of
        true ->
          Item ! {assignId, ItemId_Prod},
          NewBufferArray = BufferArray ++ [{Item, ItemType}],  % Adiciona item e tipo ao buffer
          From ! putItem,
          io:format("Produtor ~w inseriu o item ~w do tipo ~w. Há ~w itens no buffer.~n", [Prod_Id, ItemId_Prod, ItemType, length(NewBufferArray)]),
          output(NewBufferArray),
          listen(MaxSize, NewBufferArray, ItemId_Prod + 1, Type);
        false ->
          From ! full,
          listen(MaxSize, BufferArray, ItemId_Prod, Type)
      end;
    {occupied, From, Cons_Id} ->
      case BufferArray of
        [{Item, ItemType} | NewBufferArray] when length(BufferArray) > 0 ->
          ItemId_Cons = getItemId(Item),
          From ! {removeItem, ItemId_Cons, ItemType},  % Envia o tipo do item para o consumidor
          io:format("Consumidor ~w solicitou remoção do item ~w. Há ~w itens no buffer.~n", [Cons_Id, ItemId_Cons, length(NewBufferArray)]),
          output(NewBufferArray),
          listen(MaxSize, NewBufferArray, ItemId_Prod, Type);
        [] ->
          From ! empty,
          listen(MaxSize, BufferArray, ItemId_Prod, Type)
      end
  end.

output(Buffer) ->
  io:format("Buffer: ["),
  fnl(Buffer).

fnl([{Item, ItemType}]) ->
  io:format("{OP: ~w, Tipo: ~w}]~n", [getItemId(Item), ItemType]);
fnl([{Item, ItemType} | T]) ->
  io:format("{OP: ~w, Tipo: ~w}, ", [getItemId(Item), ItemType]),
  fnl(T);
fnl([]) ->
  io:format("]~n").

getItemId(Item) ->
  Item ! {self(), getId},
  receive
    {itemid, Id} ->
      Id
  end.
